fix(compiler-cli): use numeric comparison for TypeScript version

benbraou opened the pull request:
Fixes #22593

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Typescript version is compared using string compare

Issue Number: N/A
#22593

## What is the new behavior?
Typescript version is compared using numeric compare

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

benbraou commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -73,7 +74,9 @@ class AngularCompilerProgram implements Program {
+
+    if (compareVersions(ts.version, '2.4.2') < 0 ||
+        (compareVersions(ts.version, '2.7.0') >= 0 && !options.disableTypeScriptVersionCheck)) {
       throw new Error(
```
No tests were added in this PR regarding the error thrown because they are being added in https://github.com/angular/angular/pull/22701 .

benbraou commented:
Bug fix separated in 2 commits:
1. rename `symbol_query_spec.ts` to `typescript_symbols_spec.ts`
1. usage of numeric comparison instead of string comparison

By the way, diff of first commit 7d45afa89c2c7e685db450e1eb36de983e13dad6 does show that there has been a rename (however, the full PR diff shows that a file has been deleted and a new one added)

benbraou commented:
I am going to do the rebase to handle the recent merges
Also, to ease testing, I will isolate typescript version check a separate function
```typescript
export function checkTypeScriptVersion(
  version: string,
  disableTypeScriptVersionCheck: boolean | undefined
) {
  if ((version < '2.7.2' || version >= '2.8.0') && !disableTypeScriptVersionCheck) {
    throw new Error(
        `The Angular Compiler requires TypeScript >=2.7.2 and <2.8.0 but ${version} was found instead.`);
  }
}
```

vicb commented:
The first commit should be a refactor (does not "fix" anything). Thanks.

benbraou commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -858,3 +858,7 @@ function isTypescriptVersion(low: string, high?: string): boolean {
 }
+
+export function compareVersions(version1: string, version2: string): -1|0|1 {
+  return compareNumbers(toNumbers(version1), toNumbers(version2));
```
Indeed. I did not check the implementation of `compareNumbers`. Fixing + adding new tests

benbraou committed:
fix(compiler-cli): use numeric comparison for TypeScript version

Fixes #22593

benbraou commented:
@vicb Thanks for the review. PR updated
1. Utility functions that check TypeScript version are moved to a new file `typescript_version.ts`
1. `compareNumbers` method is updated
1. unit tests added for `typescript_version.ts`
1. unit test added for `checkTypeScriptVersion` in `program_spec.ts`
1. Only one commit is now needed. `symbol_query_spec.ts` is no longer renamed

vicb commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -751,6 +748,14 @@ class AngularCompilerProgram implements Program {
 
+export function checkTypeScriptVersion(
+    version: string, disableTypeScriptVersionCheck: boolean | undefined) {
+  if ((version < '2.7.2' || version >= '2.8.0') && !disableTypeScriptVersionCheck) {
```
We should probably use `compareVersions` here ?

vicb replied:
Also what about moving the version strings to const MIN_TS_VERSION and MAX_TS_VERSION ?

benbraou replied:
Indeed. I will implement a change for both comments

benbraou replied:
Indeed. I will implemented the change for both comments

vicb commented on packages/compiler-cli/test/diagnostics/typescript_version_spec.ts:
```diff
@@ -0,0 +1,74 @@
+
+describe('toNumbers', () => {
+
+  it('should handle undefined value', () => { expect(toNumbers(undefined)).toEqual([]); });
```
should it ?

benbraou replied:
- I was viewing this test as a "behavior documentation" of a method implemented prior to my commit. 
-  No, it should not handle undefined. I am changing the signature and implementation of `toNumbers`
- Great catch btw. This allowed me to discover a regression that would have been introduced. Please check my comment in `describe('isTypescriptVersion', () => {`

vicb commented on packages/compiler-cli/test/diagnostics/typescript_version_spec.ts:
```diff
@@ -0,0 +1,74 @@
+  it('should handle undefined value', () => { expect(toNumbers(undefined)).toEqual([]); });
+
+  it('should handle strings', () => {
+    expect(toNumbers('')).toEqual([]);
```
Why would this ever be valid ?

benbraou replied:
I can't think of a reason. I am removing this assertion but keeping the method implementation. Please let me know in case this is not enough

vicb reviewed (changes requested):
Mostly LGTM.
- One change required,
- Why do we need to support `toNumbers` returning an empty array ?

benbraou commented on packages/compiler-cli/test/diagnostics/typescript_version_spec.ts:
```diff
@@ -0,0 +1,74 @@
+describe('isTypescriptVersion', () => {
+  it('should correctly check if a typescript version is within a given range', () => {
+    expect(isTypeScriptVersion('2.7.0', '2.40')).toEqual(false);
+    expect(isTypeScriptVersion('2.40', '2.7.0')).toEqual(false);
```
**This assertion is incorrect.**
`2.40 > 2.7.0` => `isTypeScriptVersion('2.40', '2.7.0')` should be `true ` and not `false `
**Why is this happening ?**
```typescript
export function isTypeScriptVersion(version: string, low: string, high?: string): boolean {
const tsNumbers = toNumbers(version);
 return compareNumbers(toNumbers(low), tsNumbers) <= 0 &&
  compareNumbers(toNumbers(high), tsNumbers) >= 0;
}
```
- `isTypeScriptVersion('2.40', '2.7.0')` ( => `high = undefined`) uses internally the comparison `compareNumbers(toNumbers(undefined),[2, 40]) >= 0`
- `toNumbers(undefined)` is returning `[]` (This is counter intuitive as @vicb noticed)
- So we end up comparing [] and [2,40]
  - Prior to commit, `[] = [2,40]` because the comparison stops at the min length of both array (here 0) and defaulting to 0
  - With my commit, `[] < [2,40]` (and this makes a lot of sense for me). So, `compareNumbers(toNumbers(undefined),[2, 40] ) >= 0` returns `false` . And so,`isTypeScriptVersion('2.40', '2.7.0')` incorrectly returns `false`

**Fix ?**
- as @vicb suggested, `toNumbers(undefined)` should not even compile
- Improve the implementation of `isTypeScriptVersion`: check `high` parameter is defined before performing the comparison
- `isTypeScriptVersion` is only used in `typescript_symbols.ts` to create `SymbolTable`. Since this regression did not break any test, I am adding a unit test for `toSymbolTable` method

benbraou commented:
@vicb thanks for the review. All requested changes/questions are addressed and explained in my comments.
A regression with `isTypeScriptVersion` was fixed (I have explained my finding in my review of `typescript_version_spec.ts`)
I have added two new commits to better highlight the changes.

I see that e2e tests are failing because of this error
`Error: XHR error loading http://localhost:8000/all/benchmarks/vendor/incremental-dom-cjs.js`
I don't see how it is linked to this PR but I am having a look (maybe trigger a new build ?)

benbraou commented:
After forcing a new build, I only see `CI_MODE=saucelabs_required` failing due to timeout issues. I am seeing this same failure in other ongoing PRs

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -12,7 +12,9 @@ import * as path from 'path';
 import {BuiltinType, DeclarationKind, Definition, PipeInfo, Pipes, Signature, Span, Symbol, SymbolDeclaration, SymbolQuery, SymbolTable} from './symbols';
+import {isTypeScriptVersion} from './typescript_version';
 
+const MIN_TS_VERSION_SUPPORTING_MAP = '2.2';
```
move that right before where it is used and/or add a one line comment

benbraou replied:
will be moved and commented

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -394,21 +396,21 @@ class SignatureResultOverride implements Signature {
-      }
-      return result as ts.SymbolTable;
-    });
+export const toSymbolTable = (tsVersion: string) => (symbols: ts.Symbol[]) => {
```
Rename `toSymbolTableFactory` ?

benbraou replied:
ok

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -442,7 +444,7 @@ class SymbolTableWrapper implements SymbolTable {
     if (Array.isArray(symbols)) {
       this.symbols = symbols;
-      this.symbolTable = toSymbolTable(symbols);
+      this.symbolTable = toSymbolTable(ts.version)(symbols);
```
once we have the factory, split this line to make it more readable

benbraou replied:
ok

vicb commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,85 @@
+ * toNumbers('2.0.1'); // returns [2, 0, 1]
+ */
+export function toNumbers(value: string): number[] {
+  return value.split('.').map(v => +v);
```
nit please use `Number(v)` instead of `+v`

vicb commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,85 @@
+ */
+export function isTypeScriptVersion(version: string, low: string, high?: string): boolean {
+  const tsNumbers = toNumbers(version);
+  if (!!high) {
```
`if high !== undefined`
  and please no else after return.

`if (...) {
return
}

return;
`

benbraou commented:
@vicb I have updated the PR with a new commit to handle your latest review. Please let me know if I need to squash commits. Thanks!

rafaelss95 commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,84 @@
+ * toNumbers('2.0.1'); // returns [2, 0, 1]
+ */
+export function toNumbers(value: string): number[] {
+  return value.split('.').map(v => Number(v));
```
This can be `return value.split('.').map(Number);`

vicb replied:
I like the explicit version.

benbraou replied:
done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -394,21 +394,27 @@ class SignatureResultOverride implements Signature {
-      return result as ts.SymbolTable;
-    });
+/**
+ * Indicates the lower bound TypeScript version supporting `SymbolTable` as an `es6 Map`.
```
nit: "ES6 `Map`"

benbraou replied:
done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -394,21 +394,27 @@ class SignatureResultOverride implements Signature {
+      result[symbol.name] = symbol;
+    }
+    return result as ts.SymbolTable;
+  }
```
nits:
- add one liner comments for if cases,
- be consistent inside your fn for the castinf syntax (`as` vs `<>`),
- no `else` after `return`.

benbraou replied:
done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * Checks if a TypeScript version is greater or equal than a minimal version or in a given range of
+ * versions
```
```
* Checks if a TypeScript version is:
* - greater or equal than the provided `low` version,
* - lower or equal than an optional `high` version.
```

benbraou replied:
done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,84 @@
+ * @param low The minimum version
+ * @param high The maximum version
+ */
+export function isTypeScriptVersion(version: string, low: string, high?: string): boolean {
```
Also remove "Typescript" from the name ?

benbraou replied:
done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,84 @@
+}
+
+/**
+ * Compares two TypeScript versions
```
Remove TS

benbraou replied:
done

vicb commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -124,10 +138,7 @@ class AngularCompilerProgram implements Program {
-      throw new Error(
-          `The Angular Compiler requires TypeScript >=2.7.2 and <2.8.0 but ${ts.version} was found instead.`);
-    }
+    checkTypeScriptVersion(ts.version, options.disableTypeScriptVersionCheck);
```
make this generic (ie remove the TS part from the name) by passing MIN and MAX as paarameters ?

benbraou replied:
done
```typescript
export function checkVersion(
    version: string, minVersion: string, maxVersion: string,
    disableTypeScriptVersionCheck: boolean | undefined) {
  if ((compareVersions(version, minVersion) < 0 || compareVersions(version, maxVersion) >= 0) &&
      !disableTypeScriptVersionCheck) {
    throw new Error(
        `The Angular Compiler requires TypeScript >=${minVersion} and <${maxVersion} but ${version} was found instead.`);
  }
}
``

benbraou commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -1045,4 +1045,35 @@ describe('ng program', () => {
   });
+
+  describe('checkVersion', () => {
+    const MIN_TS_VERSION = '2.7.2';
```
Typescript min and max supported versions are intentionally defined in the test suite instead of importing them from `program.ts`. This is because `checkVersion` is generic and does not care about specific values of min and max

benbraou commented:
- PR updated based on the latest code review
- Last commit amended
- Rebase on top of master is done

vicb commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -394,21 +394,29 @@ class SignatureResultOverride implements Signature {
+  }
+
+  // ∀ Typescript version < 2.2, `SymbolTable` is implemented as a dictionary
+  const result = <any>{};
```
`const result: ts.SymbolTable = {};` and remove the cast at return time

benbraou replied:
done

vicb commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -847,6 +858,16 @@ class AngularCompilerProgram implements Program {
   }
 }
 
+export function checkVersion(
```
Add a one line comment - I think that it is important to know if the limit versions are included or not

benbraou replied:
done

vicb commented:
Please address the last comments and squash the commits ?

benbraou commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -0,0 +1,85 @@
+ * @param low The minimum version
+ * @param high The maximum version
+ */
+export function isVersion(version: string, low: string, high?: string): boolean {
```
done

benbraou commented:
Comments addressed and commits squashed.
By the way, the CI_MODE=saucelabs_required is failing (ivy tests it seems), but I see the same errors in other ongoing Pull requests.
Thanks a lot!

benbraou commented:
Build fails after rebase on top of master.
I see that https://github.com/angular/angular/pull/22997 now runs compiler-cli tests in bazel
I will update bazel config (since I renamed and added new spec files)

benbraou committed:
build(compiler-cli): include new test files in bazel config

Fixes #22593

vicb reviewed (approved):
Thanks !

vicb commented:
http://test/OCL:190874633:BASE:190874643:1522291696988:7171a71f

benbraou commented:
@vicb in `typescript_symbols.ts`, I am going to add `any` type assertion

So what we currently have:
```typescript
  // ∀ Typescript version >= 2.2, `SymbolTable` is implemented as an ES6 `Map`
    const result = new Map<string, ts.Symbol>();
    for (const symbol of symbols) {
      result.set(symbol.name, symbol);
    }
    return <ts.SymbolTable>result;
```
will become
```typescript
export const toSymbolTableFactory = (tsVersion: string) => (symbols: ts.Symbol[]) => {
  if (isVersionBetween(tsVersion, MIN_TS_VERSION_SUPPORTING_MAP)) {
    // ∀ Typescript version >= 2.2, `SymbolTable` is implemented as an ES6 `Map`
    const result = new Map<string, ts.Symbol>();
    for (const symbol of symbols) {
      result.set(symbol.name, symbol);
    }
    // First, tell the compiler that `result` is of type `any`. Then, use a second type assertion
    // to `ts.SymbolTable`.
    // Otherwise, `Map<string, ts.Symbol>` and `ts.SymbolTable` will be considered as incompatible
    // types by the compiler
    return <ts.SymbolTable>(<any>result);
  }
```

I am going to amend 6f02545ffbe87b84cfa5e4038867527eab623393

benbraou commented:
rebased on top of master again to force a new build.
Previous one has a  failing  job CI_MODE=saucelabs_required

alexeagle reviewed (approved):
bazel change looks fine to me
